docs/proposals/provisioning-proposal.md

```diff
@@ -0,0 +1,112 @@
+
+## Goals
+
+* Cluster-API provides re-usable provisioning scripts for bootstrapping a cluster for the most popular operating systems (Ubuntu, CentOS and CoreOS)
+* Provisioning scripts are well tested
+
+## Non-Goals
+
+* Implement mechanism for executing the bootstrap scripts, usable in the all scenarios. This is not subject of this proposal, but a follow-up proposal will be written soon
+
+## Implementation
+
+### Scope of Implementation
+
+The provisioning scripts are supposed to do actions described in the Proposal section: install dependencies, container runtime, Kubernetes, and configure Kubelet and Cloud Provider. Initializing Kubernetes cluster and joining nodes a cluster may or may not be part of those scripts.
+
+Cluster initialization and node joining are going to be configured using Kubeadm configuration files, which provide a big set of options for configuring cluster and all relevant cluster components (such as kube-controller-manager, kube-proxy, etc.). Cluster-API users would write kubeadm configuration files themselves and pass them as a ConfigMap/Secret or pass them to the function for generating userdata.
+
+Installing any other packages or manifests, including CNI, are not scope of those scripts, neither scope of Cluster-API. Cluster-API is not supposed to be a package manager, so everything needed should be installed as an add-on.
+
+### Implementation Prerequisites
+
+1. clusterctl should be able to install CNI. This was discussed on Cluster-API Breakout meeting held on 10/10. [Relevant issue #534](https://github.com/kubernetes-sigs/cluster-api/issues/534)
+
+### Implementation details
+
+The implementation will be based on [Go Templates](https://golang.org/pkg/text/template/). That allows implementers to source environment variables, pass data such as IP addresses, host names, and kubeadm tokens, configure Kubelet and Cloud Provider, without any change to the Cluster-API source code.
+
+Similar approach is used by many projects and providers, including:
+
+* [Cluster-API provider for DigitalOcean](https://github.com/kubermatic/cluster-api-provider-digitalocean/blob/master/cloud/digitalocean/actuators/machine/userdata.go)
+* [Cluster-API provider for GCP](https://github.com/kubernetes-sigs/cluster-api-provider-gcp/blob/master/pkg/cloud/google/metadata.go)
+* [Cluster-API provider for AWS](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/4553a80b6337b4adcc378c07db943772d30fbc78/pkg/cloud/aws/services/ec2/bastion.go)
+* [Cluster-API provider for VSphere](https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/blob/master/cloud/vsphere/provisioner/common/templates.go)
+
+The implementation will be realized using interfaces, meaning that for each supported operating system interface will be implemented. The interface has two functions, for getting user data for Master instances and for workers.
```

Comment:
@sidharthsurana: But would this mean that we are putting the go templates inside the go code? 
If that is the case, then that would make a difficult to revise these scripts when needed for a specific case. 
I suggest if we do need these pre-created scripts we store then in the form of kubernetes objects. for e.g. like a config map or even better if we can create a dedicated CRD for these. For this dedicated CRD we can publish some "standard" CRs one for each distro/version combination (like ubuntu/16.04, ubuntu/18.04, etc.). These CRs could represent the standard way of installation but at the same time if a user want to do some modifications, they can either create one more CR with their changes in. Inside the machine spec one can then refer to a particular CR of this new CRD type to specify what script to use to bootstrap the machine. This way if I am bringing my own custom image, I can also upload my custom install script in the form of a CR that I know will work on my image.

@alvaroaleman:
> But would this mean that we are putting the go templates inside the go code? 

Yes, that is something we will always have to do because there are some fields that are unique per machine, e.G. a join token or certificates.

The idea to use a CRD to make adjusting them easier is probably a good one, but I think it would be a mechanism that would then work on top of what is proposed here

@sidharthsurana:
>Yes, that is something we will always have to do because there are some fields that are unique per machine, e.G. a join token or certificates.

Well may not be necessary. The way I am thinking is this CRs for the custom CRD above would be a go template with say placeholder like "join token", "certificate" , etc. And the machine controller would read the specified CR and would mash that template with the runtime information (via the machineSpec, clusterSpec, etc.) to render the final script to be used. The point I am trying to get to is if we can avoid changing the go code everytime we want to customize that script, that would be great and make this much more useful

@roberthbailey: The GCP implementation moved towards storing these scripts in a config map (to make them easier to change). There is also the option to embed them into a bundle, which might make it easier to share them across providers.

Reply:
I find the idea of moving scripts as config maps (or better, as CRD) more appealing, because it will offer an standard operational interface (how to setup/update configuration) without making any strong assumption about the specific OS or cloud platform to be supported. IMHO this should be the main goal: define the operational interface. 

Another example of this approach, besides GCP previously mentioned is [Gardener's Operating System controller](https://github.com/gardener/gardener-extensions/tree/master/controllers/os-coreos) used to setup CoreOS nodes.

Offering a reference implementation is great, but I don't see it as the biggest benefit. This may be because In my case, I plan to use a different distribution than suggested (Ubuntu) and it is unlikely I could reuse the scripts.

Reply:
> Offering a reference implementation is great, but I don't see it as the biggest benefit.

I second this. Let's figure out the "standard operational interface (how to setup/update configuration)."